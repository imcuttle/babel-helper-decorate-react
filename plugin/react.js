const { default: createDecorateReactVisitor } = require('../')

module.exports = (babel) => {
  return {
    visitor: {
      Program(path, state) {
        const run1 = createDecorateReactVisitor({
          prefix: 'react-decorate',
          moduleInteropPath: null,
          detectScopeDepth: 1,
          ...state.opts
        })

        run1.Program.call(this, path)
      }
    }
  }
}
