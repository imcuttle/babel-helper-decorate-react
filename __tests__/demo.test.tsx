import * as babel from '@babel/core'
import * as fs from 'fs'
import * as nps from 'path'
import * as escape from 'escape-string-regexp'
import * as TestRenderer from 'react-test-renderer'
import * as React from 'react'
import * as Module from 'module'

const runCode = (code) => {
  const exports = {}
  const mod = new Module(__filename, null)
  mod.exports = exports

  const script = new Function('exports', 'module', 'require', code)
  script(exports, mod, Module.createRequire(__filename))

  return mod.exports
}

const projectRoot = nps.join(__dirname, '..')

describe('demo', () => {
  const root = nps.join(__dirname, 'lib')

  const cases = fs.readdirSync(root)
  // const cases = ['mobx']
  cases.forEach((name) => {
    // fs.readFileSync(, 'utf8')
    // fs.readFileSync(nps.join(__dirname, name, 'index.js'), 'utf8')
    // code
    const dir = nps.join(root, name, 'input')
    const names = fs.readdirSync(dir)
    names.forEach((subName) => {
      it(`demo-${name}`, () => {
        const output = babel.transformFileSync(nps.join(dir, subName), {
          presets: ['@babel/react', '@babel/env', '@babel/typescript'],
          plugins: [nps.join(root, name, 'index.js'), ['@babel/plugin-proposal-decorators', { legacy: true }]]
        }).code

        const es6Output = babel.transformFileSync(nps.join(dir, subName), {
          presets: ['@babel/typescript'],
          plugins: [nps.join(root, name, 'index.js'), '@babel/plugin-syntax-jsx']
        }).code

        const stripOutput = es6Output.replace(new RegExp(escape(projectRoot), 'g'), '<rootDir>')
        expect(stripOutput).toMatchSnapshot(`DemoEs6Code-${name}-${subName}`)

        expect(output.replace(new RegExp(escape(projectRoot), 'g'), '<rootDir>')).toMatchSnapshot(
          `DemoCode-${name}-${subName}`
        )

        if (require(nps.join(root, name, 'index.js')).runCode !== false) {
          const { Test } = runCode(output)
          const test = TestRenderer.create(<Test />)
          expect(test.toJSON()).toMatchSnapshot(`DemoOutput-${name}-${subName}`)
        }
      })
    })
  })
})
