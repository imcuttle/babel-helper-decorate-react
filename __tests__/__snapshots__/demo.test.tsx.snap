// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`demo demo-mobx: DemoCode-mobx-button.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.Test = Test;
exports[\\"default\\"] = exports.ButtonExport = void 0;

var _react = _interopRequireDefault(require(\\"<rootDir>/mobx/react.js\\"));

var React = _interopRequireWildcard(require(\\"react\\"));

var _dec, _class;

function _getRequireWildcardCache() { if (typeof WeakMap !== \\"function\\") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== \\"object\\" && typeof obj !== \\"function\\") { return { \\"default\\": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj[\\"default\\"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

function _typeof(obj) { \\"@babel/helpers - typeof\\"; if (typeof Symbol === \\"function\\" && typeof Symbol.iterator === \\"symbol\\") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === \\"function\\" && obj.constructor === Symbol && obj !== Symbol.prototype ? \\"symbol\\" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if (\\"value\\" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== \\"function\\" && superClass !== null) { throw new TypeError(\\"Super expression must either be null or a function\\"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === \\"object\\" || typeof call === \\"function\\")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError(\\"this hasn't been initialised - super() hasn't been called\\"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === \\"undefined\\" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === \\"function\\") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var Button = function Button() {
  return /*#__PURE__*/React.createElement(\\"div\\", null);
};

var ButtonExport = (0, _react[\\"default\\"])(\\"input/button.js\\")(function () {
  return /*#__PURE__*/React.createElement(\\"div\\", null);
});
exports.ButtonExport = ButtonExport;
var ButtonExportDefault = (_dec = (0, _react[\\"default\\"])(\\"input/button.js\\"), _dec(_class = /*#__PURE__*/function (_React$Component) {
  _inherits(ButtonExportDefault, _React$Component);

  var _super = _createSuper(ButtonExportDefault);

  function ButtonExportDefault() {
    _classCallCheck(this, ButtonExportDefault);

    return _super.apply(this, arguments);
  }

  _createClass(ButtonExportDefault, [{
    key: \\"render\\",
    value: function render() {
      var renderC = function renderC() {
        return /*#__PURE__*/React.createElement(\\"div\\", null);
      };

      return /*#__PURE__*/React.createElement(\\"div\\", null, renderC());
    }
  }]);

  return ButtonExportDefault;
}(React.Component)) || _class);
exports[\\"default\\"] = ButtonExportDefault;

/* mobx-decorate-disable */
function Test() {
  return /*#__PURE__*/React.createElement(\\"div\\", null, /*#__PURE__*/React.createElement(Button, null), /*#__PURE__*/React.createElement(ButtonExport, null), /*#__PURE__*/React.createElement(ButtonExportDefault, null));
}"
`;

exports[`demo demo-mobx: DemoEs6Code-mobx-button.js 1`] = `
"import _decorate from \\"<rootDir>/mobx/react.js\\";
import * as React from 'react';

const Button = () => <div></div>;

export const ButtonExport = _decorate(\\"input/button.js\\")(() => <div></div>);
export default @_decorate(\\"input/button.js\\")
class ButtonExportDefault extends React.Component {
  render() {
    const renderC = () => <div></div>;

    return <div>
      {renderC()}
    </div>;
  }

}
/* mobx-decorate-disable */

export function Test() {
  return <div>
    <Button />
    <ButtonExport />
    <ButtonExportDefault />
  </div>;
}"
`;

exports[`demo demo-mobx: DemoOutput-mobx-button.js 1`] = `
<div>
  <div />
  <div />
  <div
    mobx-data-id="input/button.js"
  >
    <div />
  </div>
</div>
`;
